File: atcoder/atc/001/b.rb
class UnionFind
    def initialize(n)
        @par = [*0...n]
        @rnk = [0]*n
    end

    def find(x)
        return x if @par[x] == x
        @par[x] = find(@par[x])
    end

    def same(x,y)
        find(x) == find(y)
    end

    def unite(x,y)
        x=find(x);y=find(y)
        return if x==y

        if @rnk[x] < @rnk[y]
            @par[x] = y
        elsif
            @par[y] = x
            @rnk[x]+=1 if @rnk[x]==@rnk[y]
        end
    end
end

n,q = gets.chomp.split.map(&:to_i)
u = UnionFind.new(n)

1.upto(q){
    p,a,b = gets.chomp.split.map(&:to_i)
    if p==0
        u.unite(a,b)
    else
        puts u.same(a,b) ? "Yes" : "No"
    end
}

File: atcoder/atc/001/b.cpp
#include <bits/stdc++.h>
using namespace std;

struct UnionFind{
    vector<int> par,rnk;

    UnionFind(int n)
    {
        par.resize(n);
        rnk.resize(n);
        for(int i=0; i<n; i++){
            par[i]=i;
            rnk[i]=0;
        }
    }

    void unite(int x,int y)
    {
        x = find(x);y = find(y);
        if(x == y) return;

        if( rnk[x] < rnk[y] )
        {
            par[x] = y;
        }
        else
        {
            par[y] = x;
            if(rnk[x] == rnk[y])
            {
                rnk[x]++;
            }
        }
    }



    bool same(int x,int y)
    {
        return find(x) == find(y);
    }

private:
    int find(int x)
    {
        if (x == par[x]) return x;
        return par[x] = find(par[x]);
    }
};

int main(){
    int N,Q;
    cin >> N >> Q;

    UnionFind u(N);

    for(int i=0; i<Q; i++){
        int p,a,b;
        cin >> p >> a >> b;

        if(p){
            cout << (u.same(a,b) ? "Yes" : "No") << endl;
        }else{
            u.unite(a,b);
        }
    }
}
